aten/src/ATen/ParallelNative.h

```diff
@@ -0,0 +1,154 @@
+#pragma once
+#include <ATen/ATen.h>
+#include <ATen/core/ivalue.h>
+
+#include <c10/core/thread_pool.h>
+
+#include <algorithm>
+#include <cstddef>
+#include <exception>
+
+#define INTRA_OP_PARALLEL
+
+namespace at {
+namespace internal {
+// internal function to get access to intra-op thread pool from
+// template parallel primitives (parallel_for, parallel_reduce)
+CAFFE2_API TaskThreadPoolBase& _get_intraop_pool();
+
+// internal utility function to mark master thread as in parallel
+// region when executing parallel primitives
+CAFFE2_API void _set_in_parallel_region(bool);
+
+// Simulate OMP's omp_get_thread_num() by force-setting thread local
+// task id as thread number when executing parallel primitives
+CAFFE2_API void _set_thread_num(size_t thread_num);
+CAFFE2_API void _unset_thread_num();
+}
+
+template <class F>
+inline void parallel_for(
+    const int64_t begin,
+    const int64_t end,
+    const int64_t grain_size,
+    const F& f) {
+  if (begin >= end) {
+    return;
+  }
+
+  if (grain_size < 0) {
+    throw std::runtime_error("Invalid begin, end or grain_size in parallel_for");
+  }
+
+  if (((end - begin) >= grain_size) && !in_parallel_region()) {
+    // choose number of tasks based on grain size and number of threads;
+    size_t chunk_size = divup((end - begin), get_num_threads());
+    // make sure each task is at least grain_size size
+    chunk_size = std::max((size_t)grain_size, chunk_size);
+    size_t num_tasks = divup((end - begin), chunk_size);
+
+    auto task = [&](int64_t task_id, int64_t local_start, int64_t local_end) {
```

Comment:
Is the capture actually necessary here? Looks like it doesn't use anything from the outer scope.

Reply:
will upd